### DISCUSSION TITLE
How to override page template output?

### DISCUSSION THREAD

**Discussion by schulle4u** [datenstrom/community#395] [Ask a question] [answered]
I'm in another round of wrestling with the API. 😀
Currently I'm redesigning my private pages plugin and would like to use a page status option like in the draft plugin. I used parts of the draft code as an example, and it mostly worked. The only thing I'm struggling with is the correct method for propper page template output. 

```
// Handle page template
    public function onParsePageTemplate($page, $name) {
        if ($this->yellow->page->get("status")=="private" && $this->yellow->getRequestHandler()=="core") {
            if (trim($_REQUEST["password"])!==$this->yellow->page->get("password")) {
                $output = "<p class=\"private-message\">".$this->yellow->page->getHtml("description")."</p>\n";
                $output .= "<form class=\"private-form\" action=\"".$this->yellow->page->getLocation(true)."\" method=\"post\">\n";
                $output .= "<p class=\"private-password\"><label for=\"password\">".$this->yellow->text->getHtml("editLoginPassword")."</label><br /><input type=\"password\" class=\"form-control\" name=\"password\" id=\"password\" value=\"".htmlspecialchars($_REQUEST["password"])."\" /></p>\n";
                $output .= "<input type=\"submit\" value=\"".$this->yellow->text->getHtml("EditLoginButton")."\" class=\"btn login-btn\" />\n";
                $output .= "</form>\n";
                $this->yellow->page->setOutput($output);
            } else {
                $this->yellow->page->set("status", "none");
            }
        }
    }
```

The password form is displayed and login works, however the page header and footer are missing from the form output. What am I doing wrong here? Should I rather use one of the content parsing options? 

Thanks
Steffen

**Comment by schulle4u** [comment]
OK, the simplest solution for now seems too respond with a page error like in the draft plugin. I'll use status code 401 in this case, which is the usual http response for unauthorized users. Works perfectly now. 
More elegant solutions are still welcome, if any.

**Comment by markseuffert** [comment, marked as answer]
Error page sounds like a really good solution for this problem, I would do the same.

Here are some alternatives. First you could override the template setting, like in the wiki plugin. This allows to create HTML output with a different template than specified in the original content page. 

Second you could request an error page, like you said. This is similar to the first alternative, it allows you to create HTML output with an error template. As side effect you'll get multi language error text.  

Third you could override page content, like in the TOC plugin. This is allows your to inject or alter HTML output that is used in templates. Useful for doing things after Markdown-to-HTML parsing.

Fourth you could override the complete template output, like in your example code above. This allows you to create output from scratch, it doesn't have to HTML output, some plugins use it for XML output.

Happy API wrestling, hope you're winning. 😄

**Comment by schulle4u** [comment]
OK, I'll use the simplest method and respond with an appropriate page error. 
Where does Yellow look for error pages in the latest core? My page-error-401.md is being deleted from `system/config` folder upon update, and the password form is no longer displayed. Using `$this->yellow->page->error("401", $pageError);` only prints out the status code, but not the error message. 

[Here is my updated code](https://github.com/schulle4u/yellow-plugins-schulle4u/tree/master/private)

Steffen

**Comment by markseuffert** [comment]
Error pages have been merged into language files, please make sure to use latest core and templates. Then you can either use `$this->yellow->page->error(500, $pageError)` which already has text with an `[yellow error]` or you can make a language file `system/plugins/language-private.txt`.

Assuming you want to use HTTP status code 403 (?), here's an example file:
```
# Datenstrom Yellow language

Language: en
Error403Title: Forbidden
Error403Text: This page is private, please enter the password. [yellow error]

Language: de
Error403Title: Verboten
Error403Text: Diese Seite ist privat, bitte gib das Kennwort ein. [yellow error]

Language: fr
Error403Title: Interdit
Error403Text: Cette page est privée, veuillez entrer le mot de passe. [yellow error]

```

**Comment by schulle4u** [comment]
Thank you. I was first assuming that multilanguage errors are also stored in the `shared` folder. Plugin fully works now!

**Comment by markseuffert** [comment]
Good redesign. I think you're using the new API faster than I can release it. Nice work.
